src/transforms/log_to_metric.rs

```diff
@@ -113,46 +115,61 @@ impl LogToMetric {
 }
 
 enum TransformError {
-    FieldNotFound,
-    TemplateError(TemplateError),
-    RenderError(String),
-    ParseError(&'static str),
+    FieldNotFound { field: Atom },
+    TemplateParseError(TemplateError),
+    TemplateRenderError { missing_keys: Vec<String> },
+    ParseFloatError { field: Atom, error: ParseFloatError },
 }
 
 fn render_template(s: &str, event: &Event) -> Result<String, TransformError> {
-    let template = Template::try_from(s).map_err(TransformError::TemplateError)?;
-    let name = template.render(&event).map_err(|e| {
-        TransformError::RenderError(format!(
-            "Keys ({:?}) do not exist on the event; dropping event.",
-            e
-        ))
-    })?;
-    Ok(String::from_utf8_lossy(&name.to_vec()).to_string())
+    let template = Template::try_from(s).map_err(TransformError::TemplateParseError)?;
+    template.render_string(&event).map_err(|missing_keys| {
+        // convert to String to avoid printing Atom in Debug format
+        let missing_keys = missing_keys
+            .into_iter()
+            .map(|k| k.to_string())
+            .collect::<Vec<String>>();
+        TransformError::TemplateRenderError { missing_keys }
+    })
 }
 
 fn render_tags(
     tags: &Option<IndexMap<Atom, String>>,
     event: &Event,
-) -> Option<BTreeMap<String, String>> {
-    match tags {
+) -> Result<Option<BTreeMap<String, String>>, TransformError> {
+    Ok(match tags {
         None => None,
         Some(tags) => {
             let mut map = BTreeMap::new();
             for (name, value) in tags {
-                if let Ok(tag) = render_template(value, event) {
-                    map.insert(name.to_string(), tag);
-                }
+                let tag = render_template(value, event)?;
+                map.insert(name.to_string(), tag);
             }
             if !map.is_empty() {
                 Some(map)
             } else {
                 None
             }
         }
-    }
+    })
 }
 
 fn to_metric(config: &MetricConfig, event: &Event) -> Result<Metric, TransformError> {
+    fn parse_field(log: &LogEvent, field: &Atom) -> Result<f64, TransformError> {
+        let value = log
+            .get(field)
+            .ok_or_else(|| TransformError::FieldNotFound {
+                field: field.clone(),
+            })?;
+        value
+            .to_string_lossy()
+            .parse()
+            .map_err(|error| TransformError::ParseFloatError {
+                field: field.clone(),
+                error,
+            })
+    }
```

Comment:
It doesn't appear this function closes over any of the parameters. Could it be moved out of `fn to_metric`?

Reply:
I moved it out!